zk-token-sdk/src/encryption/discrete_log.rs

```diff
@@ -141,35 +145,51 @@ impl DiscreteLog {
     /// Solves the discrete log problem under the assumption that the solution
     /// is a positive 32-bit number.
     pub fn decode_u32(self) -> Option<u64> {
-        let mut starting_point = self.target;
-        let handles = (0..self.num_threads)
-            .map(|i| {
-                let ristretto_iterator = RistrettoIterator::new(
-                    (starting_point, i as u64),
-                    (-(&self.step_point), self.num_threads as u64),
-                );
-
-                let handle = thread::spawn(move || {
-                    Self::decode_range(
-                        ristretto_iterator,
-                        self.range_bound,
-                        self.compression_batch_size,
-                    )
-                });
-
-                starting_point -= G;
-                handle
-            })
-            .collect::<Vec<_>>();
-
-        let mut solution = None;
-        for handle in handles {
-            let discrete_log = handle.join().unwrap();
-            if discrete_log.is_some() {
-                solution = discrete_log;
+        #[cfg(not(target_arch = "wasm32"))]
```

Comment:
I can't comment higher up, but how about changing the `num_threads` field to an `Option<NonZeroUsize>`? This has a few benefits:

* `num_threads()` can take a `NonZeroUsize` and avoid the `num_threads == 0` check
* you might need to change `new` to set `None` for wasm since it's currently defaulting to 1 otherwise
* `decode_u32` then doesn't need the `target_arch` check, and can just operate based on whether `num_threads` is set
* You can even keep the tests as they are, but just set `num_threads()` if `target_arch` isn't  wasm

What do you think?

Reply:
Oh yes, I think these are really good points. Just to make sure that I understand all the points though, I have two questions:
- What is the benefit of using `Option<NonZeroUsize>` as opposed to just `NonZeroUsize` and defaulting to `num_threads=1`? If `num_threads` field is `None`, then the `decode_u32` logic will execute assuming a single thread (if I am understanding your suggestion correctly). Then it seems like there are two representations of the same thing `Some(1)` and `None`?
- Correct me if I am wrong, but it seems like we would still need `target_arch` check in `decode_u32`, no? Since we need to remove `std::thread` entirely from the wasm logic and it seems like the compiler is not smart enough to check that `num_threads` will always be `None`.

Reply:
* Ah, I see, I was thinking `None` would mean "within the executing thread", whereas `Some(1)` would mean "in one separate thread". I guess there's no reason to do that though, is there? You're right, `1` should mean "in the same thread". In which case, there's no need for the `Option`, and it's up to you if you want to switch to a `NonZeroUsize`, but it might not be needed.
* Yes, you're right, I forgot about that part, sorry... In which case, your current implementation is probably best. I'll leave it up to you if you want to support that `num_threads = 1` behavior of executing within the same thread, but I'm not sure it's needed to be honest.

All that to say, I think what you've done is good, and probably doesn't need to be changed. But if you want `num_threads = 1` to execute in the same thread for non-wasm builds, that can be done in follow-up work.

Reply:
Ah okay, I missed the part about running in the same thread vs. in a separate thread. I think then it appears to me that an `Option<NonZeroUsize>` could be the most natural solution indeed. As you suggested, I think it makes sense to scope the PR to wasm related changes and then I can update to `Option<NonZeroUsize>` on a follow-up.